src/basic/capability-util.c

```diff
@@ -106,6 +106,14 @@ int capability_ambient_set_apply(uint64_t set, bool also_inherit) {
         _cleanup_cap_free_ cap_t caps = NULL;
         unsigned long i;
         int r;
+        uint64_t mask = 0;
+
+        /* Remove capabilities requested in ambient set, but not in the bounding set */
+        for (i = 0; i <= cap_last_cap(); i++) {
+                if ((set & (UINT64_C(1) << i)) && prctl(PR_CAPBSET_READ, i) != 1)
+                        mask |= UINT64_C(1) << i;
+        }
+        set &= ~mask;
```

Comment:
hmm, why first add the bit to a new variable `mask` just to then remove all bits set that way from `set`? I mean, you might as well drop them from `set` right-away.

also, maybe add debug logging here, to make this easier to track down:

```c
for (i = 0; i <= cap_last_cap(); i++) {
        if (!FLAGS_SET(set, UINT64_C(1) << i))
                continue;

        if (prctl(PR_CAPBSET_READ, i) > 0)
                continue;

        log_debug("Ambient capability %s requested but missing from bounding set, suppressing automatically.", capability_to_name(i));        
        set &= ~(UINT64_C(1) << i);

        if (set == 0)
                break;
}
```

or something like that?

Reply:
Agreed - I had some code printing before and after (why mask).  Updated -- hopefully that looks good, if you'd like me to rework the single if into multiple, I'm happy to do so.  I think its a bit clearer to have them in one line, but don't feel strongly.